fun.lua

```diff
@@ -590,6 +614,18 @@ local foldl = function(fun, start, gen_x, param_x, state_x)
     end
     return start
 end
+
+local funcinfo = require("jit.util").funcinfo
+
+local foldl = function(fun, start, gen_x, param_x, state_x)
+    -- Compile a separate foldl root trace for each calling function.
+    local pt = funcinfo(debug.getinfo(2, "f").func).proto
+    if foldl_clones[pt] == nil then
+        foldl_clones[pt] = clone_function(foldl_impl)
+    end
+    return foldl_clones[pt](fun, start, gen_x, param_x, state_x)
```

Comment:
Well, we've looked on this patch with @Buristan, and I believe the recipe doesn't solve the original problem described in #32.

AFAIU, the problem goes from the "bushes" of the traces generated by the function specialization. I.e. the first operand of the `map` in your example has the same semantics, but the different `GCfunc` object (i.e. function address, anchored in the assertion guard) and even `GCproto`.

However, I see that you try to "cache" or "linearize" another function, that is a `reduce` caller. As a result, we see no difference for your changes in the `jit.dump` output.

I guess, that the `fun` argument should be "cached", if one tries to avoid generation of side traces for different function objects with the same prototypes. For the case of the same "literal prototypes" implemented by different function prototypes, the situation goes much more deeper.